Rewrite of DPT implementation

DocLambda opened the pull request:
When trying to instantiate datapoints from group-address to DPT mappings declared in a config file, new instances of datapoints need to be created from DPT names like "9.001" or similar. 

Therefore, the proposed changes first add some meta-data like unit or type of measurement to the datapoint. Afterwards, a necessary getter function for getting the actual value of the datapoint is provided and added to the interface. Eventually, a function following the factory pattern is implemented to provide structure instances from given standard KNX DPT names.

vapourismo commented:
This strikes me as very odd: Why are the datapoint *type* information encoded at the *value* level?
From a library perspective, this kind of information is not relevant.

I agree, the names for DPT types are a bit odd because it follows the colloquial names from the specification and not the numbered ones.

I'd much rather see a complete array of:
  1. `unpackX` and `packX` helpers for each bit pattern
  2. Types corresponding to the numbered DPTs (X.YYY) making use of the underlying `packX`/`unpackX` function but keeping the representation of the particular DPT

DocLambda commented:
I see your point. I guess the problem is that I wanted to tackle two things with this PR. The target is to

1. Add new datapoint types with minimal code duplication.
2. Make the semantics of the datapoints (what is represented) machine readable.

Let me detail the points. Reading the KNX spec, given a DPT X.YYY, the data **format** is the same within one X that is e.g. all DPT 9.YYY are represented the same way on the wire and thus would share the same `pack` and `unpack` functions. However, all YYY of the 9.YYYY have different semantics that is the measured variable and the unit differ e.g. 9.001 is a "temperature" with unit "degree Celsius" and 9.001 is an "illumination" with unit "lux" etc.
So the idea was to separate the datapoint **format** and the **semantics** and make both accessible to users of the library. This is also touching the second point to avoid that every user of the library has to look-up and implement the mapping between format and semantics on her own.

So how about the following:
We implement one type for each DPT X and provide the corresponding `pack` and `unpack` functions similar to what you had before. Additionally, we implement a function (e.g. `GetDatapointTypeInfo()`) to get the semantic information (measured variable and unit) of a datapoint type.

What do you think? If this is ok for you or you have some other proposal, please let me know and I will cook up something...

DocLambda commented:
After some tests allow me to ask some more questions:

1. The strict typing of go does not allow to share the `Pack` and `Unpack` functions. This is also where your current code fails. Try to use an `OpenClose` as `DatapointValue`... So the only way is to implement all interface functions (`Pack`, `Unpack`, `String`) for each subtype `DTP_YXXX` for all `XXX` where e.g. all DPT 1.xxx and DPT 9.xxx share the exact same `Pack` and `Unpack` function. Is this what you want? Just to make sure....
2. Should I submit a separate pull request to fix the DPT 1.xxx implementation?
2.  In case the answer to point 1 is 'yes', do you accept a code generator generating the above code to reduce the monkey work to a minimum?
3. Is it ok to add a `GetUnit` function to the `DatapointValue` as this IMO belongs to the value?
4. Will you accept a `Produce` function to get a `DatapointValue` for a given string (e.g. "1.001")?

Thanks for your time and advice!

vapourismo commented:
1. Types representing datapoint types in the group DPT X do not share the same `Pack` and `Unpack` functions. That's why they are separate types.

Consider the example I commented previously: DPT 9.001 and 9.004.

9.001 has a lower bound of -273 Degree Celsius, while 9.004 has a lower bound of 0 Lux.
The `Unpack` method for these types need to reflect these properties.

Types give us a neat way to encode unit, resolution and bounds statically.

2. Can you elaborate what's wrong with it? Maybe explain in a seperate PR if the implementation is really broken.

3. I guess that's Go for you. The language just isn't very good for this kind of DRY stuff, or anything really. So I suppose my answer would be yes as I'd be interested to see how one does it. 

4. A seperate interface that demotes meta data such as unit, resolution, bounds or DPT ID to the value level is perfectly valid - as previously stated.

5. A function `func MakeDatapointValue(id DatapointTypeID, value interace{}) (DatapointValue, error)` would be nice. I just don't like the stringly-typedness.

DocLambda commented:
Here comes the new trial. Sorry for resetting the branch... :-S

1.  Understood.
2. A bug report will follow (but probably next year...) ;-)
3. Main part of this series (see below)
4. IMO the unit belongs to a value as otherwise it's just a number. That's why you also add it in your `String()` implementation...
5. This won't solve my problem. I get a string from a config file (e.g. "9.001") for a group-address (e.g. "5/5/1"). Now I need to instantiate a `DatapointValue` with the correct type to parse the value when receiving a message for "5/5/1". So with your interface I would need to implement a LUT to derive the 
`DatapointTypeID` for my string just to push it to another LUT to derive the `DatapointValue` for the 
`DatapointTypeID`. Instead I might directly implement a LUT for directly getting the `DatapointValue` from the string on my side. It's just that everybody with such a config will need to do the same...

Details on the commits:

The first two commits add the template and generator as well as the DPT definition file. 
Use `go run generator -template types.tmpl > types.go` to generate the source file.
The third commit will remove the existing types.go, so please skip it if you want to play around...

Commit four adds a unit test for doing a pack/unpack cycle to check if the implementation is good.
The following commits show how to implement new DPTs. Please note, after adding DPTs to the json file you need to regenerate the `types.go` and `types_test.go` files.

vapourismo commented:
4. The unit in the String method is static, as it is supposed to be. If you make it part of the value, it becomes dynamic. Anyone can change that value and therefore change the unit, resolution or range of that particular value, rendering it inaccurate. 

Consider this instead (and analogously for other static properties):

```go
function (DPT9004) Unit() { 
    return "Lux"
}
```

5. That is hardly a problem that this library should solve. Your parser should arguably exist in your package, not this library.

On a side note, please use `gofmt`.

DocLambda commented:
4. Perfectly fine with me. Will implement the `Unit` function you propose.
5. Ok. I will solve this problem on my side then...

Do you agree with the approach and the series in general? If so, I will send an update with correct formatting (sorry, I'm new to go) and the `Unit` function above.
Are there any further points before you will accept the pull request?

vapourismo commented:
Yes, I am fine with the discussed approach.

DocLambda commented:
In this revision:

- Fixed formatting so gofmt does not complain. Well I cannot get rid of the double newline at the end of the generated types.go and types_test.go, but I hope this is acceptable.
- Added Unit() function according to your suggestion. If there is no unit the function returns `nil`
- Added cosmetic whitespace in the String() function to better separate value and unit

DocLambda commented:
Return an empty string if no unit exists for a given type as we cannot return `nil`.

vapourismo commented:
Would be nice if the generated file is also checked in. Ideally the CI should perform a check whether the current version and the would-be version have diverged.

DocLambda commented:
Added the generated files.
I have no idea how to add the test you propose to TravisCI, but I think it makes a lot of sense...

vapourismo commented on knx/types_test.go:
```diff
@@ -0,0 +1,263 @@
+// Copyright 2017 Ole Krüger.
+// Licensed under the MIT license which can be found in the LICENSE file.
+
+package dpt
```
This is the wrong package. This file should be in `knx/dpt`.

DocLambda commented:
Fixed the template and moved the generated unit-test file to the correct location.
